File: server/service/miner.py
import nmap
import requests
import sqlite3

PORTS = "3333,4444,5555,7777,8888,18081,18082,18083,4028,4233,8233"
DB = "miners.db"

def scan(targets):
    nm = nmap.PortScanner()
    nm.scan(hosts=targets, ports=PORTS, arguments="-sT -Pn --open")
    conn = sqlite3.connect(DB)
    for host in nm.all_hosts():
        open_ports = []
        for proto in nm[host].all_protocols():
            for port in nm[host][proto].keys():
                open_ports.append(port)
        if open_ports:
            geo = requests.get(f"http://ip-api.com/json/{host}").json()
            conn.execute(
                "INSERT OR IGNORE INTO miners (ip, open_ports, lat, lon, city, country) VALUES (?, ?, ?, ?, ?, ?)",
                (host, ",".join(map(str, open_ports)), geo.get("lat"), geo.get("lon"), geo.get("city"), geo.get("countryCode")),
            )
    conn.commit()
    conn.close()

File: server/service/miner.ts
import * as fs from "fs";
import * as ExcelJS from "exceljs";
export async function exportMiners(format: string, filter: string) {
  const miners = /* fetch filtered miners from DB */;
  if (format === "csv") {
    const csv = miners.map(m => `${m.ip},${m.open_ports},${m.city},${m.country}`).join("\n");
    fs.writeFileSync("/tmp/miners.csv", csv);
    return "/tmp/miners.csv";
  }
  if (format === "xlsx") {
    const wb = new ExcelJS.Workbook();
    // ... fill workbook
    await wb.xlsx.writeFile("/tmp/miners.xlsx");
    return "/tmp/miners.xlsx";
  }
  if (format === "json") {
    fs.writeFileSync("/tmp/miners.json", JSON.stringify(miners));
    return "/tmp/miners.json";
  }
}